pkg/replication/metrics.go

```diff
@@ -0,0 +1,105 @@
+		Name:    "immudb_replication_commit_time",
+		Buckets: prometheus.ExponentialBucketsRange(0.001, 10.0, 16),
+		Help:    "histogram of time spent by replicators to replicate a single transaction",
+	}, []string{"db"})
+
+	_metricsReplicators = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "immudb_replication_replicators",
+		Help: "number of replicators available",
+	}, []string{"db"})
+
+	_metricsReplicatorsActive = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "immudb_replication_replicators_active",
+		Help: "number of replicators actively processing transactions",
+	}, []string{"db"})
+
+	_metricsReplicatorsInRetryDelay = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "immudb_replication_replicators_retry_delay",
+		Help: "number of replicators that are currently delaying the replication due to a replication error",
+	}, []string{"db"})
+
+	_metricsReplicationRetries = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "immudb_replication_replicators_retries",
+		Help: "number of retries while replicating transactions caused by errors",
+	}, []string{"db"})
+
+	_metricsReplicationPrimaryCommittedTxID = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "immudb_replication_primary_committed_tx_id",
+		Help: "the latest know transaction ID committed on the primary node",
+	}, []string{"db"})
+
+	_metricsAllowCommitUpToTxID = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "immudb_replication_allow_commit_up_to_tx_id",
+		Help: "most recently received confirmation up to which commit id the replica is allowed to durably commit",
+	}, []string{"db"})
+)
+
+type metrics struct {
+	txWaitQueueHistogram     prometheus.Observer
+	replicationTimeHistogram prometheus.Observer
+	replicationRetries       prometheus.Counter
+	replicators              prometheus.Gauge
+	replicatorsActive        prometheus.Gauge
+	replicatorsInRetryDelay  prometheus.Gauge
+	primaryCommittedTxID     prometheus.Gauge
+	allowCommitUpToTxID      prometheus.Gauge
+}
+
+// metricsForDb returns metrics object for particular database name
+func metricsForDb(dbName string) metrics {
+	return metrics{
+		txWaitQueueHistogram:     _metricsTxWaitQueueHistogram.WithLabelValues(dbName),
+		replicationTimeHistogram: _metricsReplicationTimeHistogram.WithLabelValues(dbName),
+		replicationRetries:       _metricsReplicationRetries.WithLabelValues(dbName),
+		replicators:              _metricsReplicators.WithLabelValues(dbName),
+		replicatorsActive:        _metricsReplicatorsActive.WithLabelValues(dbName),
+		replicatorsInRetryDelay:  _metricsReplicatorsInRetryDelay.WithLabelValues(dbName),
+		primaryCommittedTxID:     _metricsReplicationPrimaryCommittedTxID.WithLabelValues(dbName),
+		allowCommitUpToTxID:      _metricsAllowCommitUpToTxID.WithLabelValues(dbName),
+	}
+}
+
+// reset ensures all necessary gauges are zeroed
+func (m *metrics) reset() {
+	m.replicators.Set(0)
+	m.replicatorsActive.Set(0)
+	m.replicatorsInRetryDelay.Set(0)
```

Comment:
primaryCommittedTxID and allowCommitUpToTxID do not need to be reset?

Reply:
My assumption was false about the gauge - once it's instantiated for given db it starts appearing in prometheus output. I've added reset to 0 for missing metrics